molecules/pubchem/pipeline.py

```diff
@@ -0,0 +1,193 @@
+          'atom_symbol': line[31:34].strip(),
+          'mass_difference': int(line[34:36]),
+          'charge': int(line[36:39]),
+          'atom_stereo_parity': int(line[39:42]),
+          'hydrogen_count': int(line[42:45]),
+          'stereo_care_box': int(line[45:48]),
+          'valence': int(line[48:51]),
+          'h0_designator': int(line[51:54]),
+          'atom_atom_mapping_number': int(line[60:63]),
+          'inversion_retention': int(line[63:66]),
+          'exact_change_flag': int(line[66:69]),
+        })
+
+      bonds = []
+      start = end
+      end = start + total_bonds
+      for line in molecule[sdf.MDF_SECTION][start:end]:
+        bonds.append({
+          'first_atom_number': int(line[0:3]),
+          'second_atom_number': int(line[3:6]),
+          'bond_type': int(line[6:9]),
+          'bond_stereo': int(line[9:12]),
+          'bond_topology': int(line[15:18]),
+          'reacting_center_status': int(line[18:21]),
+        })
+
+      molecule.update({
+          'is_chiral': is_chiral,
+          'ctab_version': ctab_version,
+          'atoms': atoms,
+          'bonds': bonds,
+      })
+      molecule.pop(sdf.MDF_SECTION)
+      yield molecule
+
+    except Exception:
+      logging.exception(pprint.pformat(molecule))
+
+
+class CountAtoms(beam.DoFn):
+  def count_by_atom_symbol(self, molecule, atom_symbol):
+    # For every atom in the molecule, check if the atom type equals the
+    # element we're trying to count. If True, it will be casted to a 1,
+    # otherwise it will be a 0, then sum all of them.
+    return sum(int(atom['atom_symbol'] == atom_symbol)
+        for atom in molecule['atoms'])
+
+  def process(self, molecule):
+    uid = int(molecule['<PUBCHEM_COMPOUND_CID>'][0])
+    label = float(molecule['<PUBCHEM_MMFF94_ENERGY>'][0])
+    result = {
+      'ID': uid,
+      'TotalC': self.count_by_atom_symbol(molecule, 'C'),
+      'TotalH': self.count_by_atom_symbol(molecule, 'H'),
+      'TotalO': self.count_by_atom_symbol(molecule, 'O'),
+      'TotalN': self.count_by_atom_symbol(molecule, 'N'),
+      'Energy': label,
+    }
+    yield result
+
+
+class SimpleFeatureExtraction(beam.PTransform):
+  """The feature extraction (element-wise transformations).
+
+  We create a `PTransform` class. This `PTransform` is a bundle of
+  transformations that cann be applied to any other pipeline as a step.
+
+  We'll extract all the raw features here. Due to the nature of `PTransform`s,
+  we can only do element-wise transformations here. Anything that requires a
+  full-pass of the data (such as feature scaling) has to be done with
+  tf.Transform.
+  """
+  def __init__(self, source):
+    super(SimpleFeatureExtraction, self).__init__()
+    self.source = source
+
+  def expand(self, p):
+    # Return the preprocessing pipeline. In this case we're reading the PubChem
+    # files, but the source could be any Apache Beam source.
+    return (p
+        | 'Read from source' >> self.source
```

Comment:
It may be just a technicality, but here, `self.source` is a PTransform rather than a source. It is a `beam.Read(...)` PTransform that reads from a source (source may be a PubSubSource, or something else).
We may want to rename to `read_transform` or something like that, just for clarity. I leave it up to you.

Reply:
Done